cdc/log.cc

```diff
@@ -570,23 +595,141 @@ class transformer final {
+                                        process_cells([&](const bytes_view& key, const bytes_view& value, bool live) {
+                                            result.emplace_back(key);
+                                        });
+                                        return type.serialize_partially_deserialized_form(result, cql_serialization_format::internal());
+                                    },
+                                    // for user type we collect the fields in the mutation and set to
+                                    // tuple of value or tuple of null in case of delete.
+                                    // fields not in the mutation are null in the enclosing tuple, signifying "no info"
+                                    [&](const user_type_impl& type) {
+                                        std::vector<bytes_opt> res(type.size());
+                                        std::array<bytes_view_opt, 1> tmp;
+                                        // type of the actual value in log column tuple.
+                                        auto res_type = static_pointer_cast<const tuple_type_impl>(dst->type)->type(1);
+                                        auto tt = static_pointer_cast<const tuple_type_impl>(res_type);
+                                        process_cells([&](const bytes_view& key, const bytes_view& value, bool live) {
+                                            auto idx = deserialize_field_index(key);
+                                            tmp[0] = live ? bytes_view_opt{value} : std::nullopt;
+                                            res[idx].emplace(static_pointer_cast<const tuple_type_impl>(tt->all_types()[idx])->build_value(tmp));
+                                        });
+                                        return tt->build_value(res);
+                                    },
+                                    [&] (const abstract_type& o) -> bytes {
+                                        throw std::runtime_error(format("cdc transform: unknown type {}", o.name()));
+                                    }
+                                ));
+                            });
+
+                            op = oop.value_or(op);
+                        }
+
+                        values[0] = data_type_for<column_op_native_type>()->decompose(data_value(static_cast<column_op_native_type>(op)));
+                        res.set_cell(log_ck, *dst, atomic_cell::make_live(*dst->type, ts, tuple_type_impl::build_value(values)));
+
+                        if (has_pirow) {
+                            values[0] = data_type_for<column_op_native_type>()->decompose(data_value(static_cast<column_op_native_type>(column_op::set)));
+                            values[1] = cdef.is_atomic()
+                                ? pirow->get_blob(cdef.name_as_text())
+                                : values[1] = visit(*cdef.type, make_visitor(
```

Comment:
Why do we need to do values[1] = visit instead of just visit here? Is this needed because ?: requires same types?